ec/src/models/mod.rs

```diff
@@ -9,16 +9,29 @@ pub mod mnt6;
 pub mod short_weierstrass_jacobian;
 pub mod twisted_edwards_extended;
 
+/// Model parameters for an elliptic curve.
 pub trait ModelParameters: Send + Sync + 'static {
     type BaseField: Field + SquareRootField;
     type ScalarField: PrimeField + SquareRootField + Into<<Self::ScalarField as PrimeField>::BigInt>;
+    type Affine: AffineCurve;
+
+    const COFACTOR: &'static [u64];
+    const COFACTOR_INV: Self::ScalarField;
+
+    /// Checks that the current point is in the prime order subgroup given
+    /// the point on the curve.
+    /// Requires type parameters G: the type of point passed in, and H: the type of
+    /// point that results from multiplying G by a scalar.
+    fn is_in_correct_subgroup_assuming_on_curve(item: &Self::Affine) -> bool {
+        item.mul_bits(BitIteratorBE::new(Self::ScalarField::characteristic()))
+            .is_zero()
+    }
 }
 
+/// Model parameters for a Short Weierstrass curve.
 pub trait SWModelParameters: ModelParameters {
```

Comment:
```suggestion
pub trait SWModelParameters: ModelParameters<Affine = short_weierstrass_jacobian::GroupAffine> {
```

Reply:
It gives me
```
the size for values of type `Self` cannot be known at compilation time
```

Reply:
Huh adding `+ Sized` to `ModelParameters` seems to fix it